src/mpl/src/gpu/mpl_gpu_cuda.c

```diff
@@ -395,7 +400,7 @@ int MPL_gpu_finalize(void)
     }
 
     /* Reset initialization state */
-    gpu_initialized = 0;
+    MPL_atomic_store_int(&gpu_initialized, MPL_CUDA_UNINIT);
```

Comment:
One concern is that just calling `cudaFree` in the application will now initialize mpl/cuda. We can skip the hook logic if it's not initialized, but we still need to call the underlying `cudaFree` no matter what. Maybe it would be better to only initialize the function pointers?

Reply:
Actually, I think we need to go further and make the memory hooks completely thread-safe with a lock. Otherwise we are breaking the interface provided by CUDA. I'll update the PR.